internal/controllers/reference_addon_controller.go

```diff
@@ -2,44 +2,110 @@ package controllers
 
 import (
 	"context"
-	"strings"
+	"fmt"
 
 	"github.com/go-logr/logr"
-	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 
 	refapisv1alpha1 "github.com/openshift/reference-addon/apis/reference/v1alpha1"
+	"github.com/openshift/reference-addon/internal/controllers/phase"
 	"github.com/openshift/reference-addon/internal/metrics"
+	corev1 "k8s.io/api/core/v1"
 )
 
-type ReferenceAddonReconciler struct {
-	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
-}
-
-func (r *ReferenceAddonReconciler) Reconcile(
-	ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	log := r.Log.WithValues("addon", req.NamespacedName.String())
-
-	// dummy code to indicate reconciliation of the reference-addon object
-	if strings.HasPrefix(req.NamespacedName.Name, "redhat-") {
-		log.Info("reconciling for a reference addon object prefixed by redhat- ")
-	} else if strings.HasPrefix(req.NamespacedName.Name, "reference-addon") {
-		log.Info("reconciling for a reference addon object named reference-addon")
-	} else {
-		log.Info("reconciling for a reference addon object not prefixed by redhat- or named reference-addon")
+func NewReferenceAddonReconciler(client client.Client, opts ...ReferenceAddonReconcilerOption) (*ReferenceAddonReconciler, error) {
+	var cfg ReferenceAddonReconcilerConfig
+
+	cfg.Option(opts...)
+	cfg.Default()
+
+	signaler, err := NewConfigMapUninstallSignaler(
+		client,
+		WithAddonNamespace(cfg.AddonNamespace),
+		WithOperatorName(cfg.OperatorName),
+		WithDeleteLabel(cfg.DeleteLabel),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("initializing uninstall signaler: %w", err)
 	}
 
-	// add sample metrics
-	metrics.RequestSampleResponseData()
+	return &ReferenceAddonReconciler{
+		cfg: cfg,
+		phases: []phase.Phase{
+			NewPhaseUninstall(
```

Comment:
a comment mentioning that the order of phases matter (or is recommended) will be helpful for other developers. As the Uninstall phase is intented to remove resources and block (skip) further Reconcile phases.